[FIX][QAT] Switch to use `kwargs` when `args` is empty

sijiac opened the pull request:
Summary:
When `ref_node.args` is empty, the QAT will throw index out of range. Here is an example, line 574 is using `tensors = ....` in torch.cat func, which will be treated as `kwargs`
{F800357376}

f388506954

To fix the issue, we will use the value of the first kwarg if args is empty

Test Plan: f388545532

Reviewed By: bigning, lyoka

Differential Revision: D41396771



cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

houseroad reviewed (approved):
Looks good.

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -829,7 +829,8 @@ def special_pattern_replacement(model: QuantizedGraphModule):
         if not (is_call_module or is_call_function or is_call_method):
             continue
-        dq_node_or_nodes = ref_node.args[0]
+        assert len(ref_node.args) > 0 or len(ref_node.kwargs) > 0
```
if it works for your case, maybe we can start with requiring len(ref_node.kwargs) == 1

sijiac replied:
for `torch.cat`, there would be two kwargs, `tensors` and `dim`. It may not work if we limit the number to be 1

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -829,7 +829,8 @@ def special_pattern_replacement(model: QuantizedGraphModule):
             continue
-        dq_node_or_nodes = ref_node.args[0]
+        assert len(ref_node.args) > 0 or len(ref_node.kwargs) > 0
+        dq_node_or_nodes = ref_node.args[0] if len(ref_node.args) > 0 else list(ref_node.kwargs.items())[0][1]
```
nit: `list(ref_node.kwargs.items())[0][1]` --> `list(ref_node.kwargs.values())[0]`

sijiac committed:
[FIX][QAT] Switch to use `kwargs` when `args` is empty (#89778)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/89778

When `ref_node.args` is empty, the QAT will throw index out of range. Here is an example, line 574 is using `tensors = ....` in torch.cat func, which will be treated as `kwargs`
{F800357376}

f388506954

To fix the issue, we will use the value of the first kwarg if args is empty

Test Plan: f388545532

Reviewed By: bigning, lyoka

Differential Revision: D41396771

fbshipit-source-id: 40b884618b96903565aa8577f2f8a363f8265446
